netaddr.go

```diff
@@ -412,96 +361,76 @@ func (ip IP) Is6() bool {
 // address it's returned unchanged.
 func (ip IP) WithZone(zone string) IP {
-	if zone == "" {
-		if z, ok := ip.ipImpl.(v6AddrZone); ok {
-			return IP{z.v6Addr}
-		}
+	if !ip.Is6() {
 		return ip
 	}
-	switch ip := ip.ipImpl.(type) {
-	case v6Addr:
-		return IP{v6AddrZone{ip, zone}}
-	case v6AddrZone:
-		return IP{v6AddrZone{ip.v6Addr, zone}}
+	if zone == "" {
+		ip.zt = z6zoneless
+		return ip
 	}
+	panic("XXX TODO allocate one or get from table")
 	return ip
 }
 
 // IsLinkLocalUnicast reports whether ip is a link-local unicast address.
 // If ip is the zero value, it will return false.
 func (ip IP) IsLinkLocalUnicast() bool {
-	// See: https://en.wikipedia.org/wiki/Link-local_address.
-	switch ip := ip.ipImpl.(type) {
-	case nil:
-		return false
-	case v4Addr:
-		return ip[0] == 169 && ip[1] == 254
-	case v6Addr:
-		return ip[0] == 0xfe && ip[1] == 0x80
-	case v6AddrZone:
-		return ip.v6Addr[0] == 0xfe && ip.v6Addr[1] == 0x80
-	default:
-		panic("netaddr: unhandled ipImpl representation")
+	if ip.Is4() {
+		return ip.a[o4+0] == 169 && ip.a[o4+1] == 254
+	}
+	if ip.Is6() {
+		return ip.a[0] == 0xfe && ip.a[1] == 0x80
 	}
+	return false
 }
 
 // IsLoopback reports whether ip is a loopback address. If ip is the zero value,
 // it will return false.
 func (ip IP) IsLoopback() bool {
-	switch ip := ip.ipImpl.(type) {
-	case nil:
-		return false
-	case v4Addr:
-		return ip[0] == 127
-	case v6Addr:
-		return string(ip[:len(v6Loopback)]) == v6Loopback
-	case v6AddrZone:
-		return string(ip.v6Addr[:len(v6Loopback)]) == v6Loopback
-	default:
-		panic("netaddr: unhandled ipImpl representation")
+	if ip.Is4() {
+		return ip.a[o4+0] == 127
 	}
+	if ip.Is6() {
+		return string(ip.a) == v6Loopback
+	}
+	return false
 }
 
 // IsMulticast reports whether ip is a multicast address. If ip is the zero
 // value, it will return false.
 func (ip IP) IsMulticast() bool {
-	// See: https://en.wikipedia.org/wiki/Multicast_address.
-	switch ip := ip.ipImpl.(type) {
-	case nil:
-		return false
-	case v4Addr:
-		return ip[0]&0xf0 == 0xe0
-	case v6Addr:
-		return ip[0] == 0xff
-	case v6AddrZone:
-		return ip.v6Addr[0] == 0xff
-	default:
-		panic("netaddr: unhandled ipImpl representation")
+	if ip.Is4() {
+		return ip.a[o4+0]&0xf0 == 0xe0
+	}
+	if ip.Is6() {
+		return ip.a[0] == 0xff
 	}
+	return false
 }
 
 // Prefix applies a CIDR mask of leading bits to IP, producing an IPPrefix
 // of the specified length. If IP is the zero value, a zero-value IPPrefix and
 // a nil error are returned. If bits is larger than 32 for an IPv4 address or
 // 128 for an IPv6 address, an error is returned.
 func (ip IP) Prefix(bits uint8) (IPPrefix, error) {
-	if ip.ipImpl == nil {
+	if ip.zt == z0 {
 		return IPPrefix{}, nil
 	}
-	// TODO: optimize this to return ip directly if it's already
-	// masked and to not allocate a new IP.ipImpl.
-	return ip.prefix(bits)
+	if ip.Is4() {
+		return v4Prefix(ip.As4(), bits)
+	}
+	// TODO: preserve zone?
```

Comment:
we should delegate that decision to @mdlayher, as corerad is the most zone-aware thing we know of using netaddr. Matt, does zone in a prefix make any sense to have?

Barring a known use case, I'm inclined to keep the zone anyway, because it can be used to e.g. mean "the link-local fe80:: prefix on a specific interface", although I can't think how I'd use it for same.

Reply:
I can't think of any reason you'd need zone on a prefix. I don't think I've ever seen eg "fe80::%eth0/64" or similar. But I'll think on it.

Reply:
Actually we had tests that it was preserved (that I failed), so I put it back.